src/verbs/device_context.rs

```diff
@@ -142,4 +150,89 @@ impl DeviceContext {
             }
         }
     }
+
+    pub fn legacy_query_gid_table(&self) -> Result<Vec<GidEntry>, String> {
+        let mut res = Vec::new();
+        let dev_attr = self.query_device().unwrap();
+        let mut gid_type;
+
+        for port_num in 1..(dev_attr.phys_port_cnt() + 1) {
+            let port_attr = self.query_port(port_num).unwrap();
+
+            if let Some(name) = self.name() {
+                for gid_index in 0..port_attr.gid_tbl_len() {
+                    let gid = self.query_gid(port_num, gid_index).unwrap();
+                    let netdev_index;
+
+                    if gid.is_zero() {
+                        continue;
+                    }
+
+                    unsafe {
+                        gid_type = match self.query_gid_type(port_num, gid_index as u32).unwrap_unchecked() {
+                            0 if port_attr.link_layer() == IBV_LINK_LAYER_INFINIBAND => 0,
+                            0 if port_attr.link_layer() == IBV_LINK_LAYER_ETHERNET => 1,
+                            1 => 2,
+                            num => panic!("unknown gid type {num}!"),
+                        };
+                    }
+
+                    let netdev = unsafe {
+                        fs::read_to_string(format!(
+                            "/sys/class/infiniband/{}/ports/{}/gid_attrs/ndevs/{}",
+                            name, port_num, gid_index
+                        ))
+                        .unwrap_unchecked()
+                        .trim_ascii_end()
+                        .to_string()
+                    };
+
+                    unsafe {
+                        netdev_index = libc::if_nametoindex(CString::new(netdev).unwrap_unchecked().as_ptr());
+                    }
+
+                    res.push(GidEntry(ibv_gid_entry {
+                        gid: gid.into(),
+                        gid_index: gid_index as u32,
+                        port_num: port_num.into(),
+                        gid_type,
+                        ndev_ifindex: netdev_index,
+                    }))
+                }
+            } else {
+                return Err(format!("device doesn't have a valid name"));
+            }
+        }
+
+        Ok(res)
+    }
+
+    #[inline]
+    pub fn query_gid_table(&self) -> Result<Vec<GidEntry>, String> {
+        let mut entries = vec![GidEntry::default(); 32];
+        let mut size = 0;
+
+        unsafe {
+            size = _ibv_query_gid_table(self.context, entries.as_mut_ptr() as _, 32, 0, 32);
```

Comment:
Maybe in the future we should introduce the definition of `ibv_query_gid_table` to rdma-mummy-sys rather than invoking the internal function with leading underscore directly?

Reply:
Yes, but the wrapper should be implemented by us manually, as it's an inline function which just add a `sizeof()` params into `_ibv_query_gid_table`

```
static inline ssize_t ibv_query_gid_table(struct ibv_context *context,
					  struct ibv_gid_entry *entries,
					  size_t max_entries, uint32_t flags)
{
	return _ibv_query_gid_table(context, entries, max_entries, flags,
				    sizeof(*entries));
}
```
For Rust Vec or Slices, we need to pass `vec.size()`

Reply:
I see. I think we can refine the implementation of rdma-mummy-sys later. I will open some issues at rdma-mummy-sys to keep track of these stuffs.

Reply:
Updated with calculated vector size for fitting in all possible valid GIDs.